aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -22,12 +22,41 @@ using namespace at::sparse;
 
 namespace {
 
-  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input) {
-    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true)) {
+  inline SparseTensor get_result_tensor_for_unary_op(const SparseTensor& input, bool promotes_integer_to_float = false) {
+    if (c10::isIntegralType(input.scalar_type(), /*includeBool=*/true) && promotes_integer_to_float) {
       return at::empty_like(input, input.options().dtype(c10::get_default_dtype()));
     }
     return at::empty_like(input);
   }
+
+  // --------------------------------------------------------------------
+  // helper function for sanity check of input and output tensors
+  // this helper function is for unary ops
+  // --------------------------------------------------------------------
+  static inline void sparse_unary_op_check_sanity(SparseTensor& result, const SparseTensor& self,
+      bool promotes_integer_to_float = false, bool is_linear_map = false) {
+    TORCH_CHECK(result.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.is_sparse(), "Tensor should be sparse");
+    TORCH_CHECK(self.get_device() == result.get_device(),
+        "Both tensors should be on the same device");
+
+    if (promotes_integer_to_float){
+      TORCH_CHECK(
+        !c10::isIntegralType(result.scalar_type(), /*includeBool=*/true),
+        "result type cannot be Integral, got:",
+        result.scalar_type());
+    }
+
+    if (is_same_tensor(result, self) && is_linear_map) {
+      // don't have in-place asin for uncoalesced input because coalesce() is not
+      // in-place, see above comment
```

Comment:
The `is_same_tensor` does not appear to be needed. Here's my suggestion:
```
if (is_linear_map) {  // self is allowed to be uncoaleased
  copy_sparse_to_sparse_(result, self);  // copies only when result and self are not the same, result can be uncoaleased
} else {
  copy_sparse_to_sparse_(result, self.coalesce());  // resizes result and copies coalesced self to it
}
```

}
```

Reply:
@pearu - made the changes as suggested